### PR TITLE
add patch behavior

### PR DESCRIPTION
**Pull request by dmacthedestroyer**
this addresses #379 

If a binding is mapped to `patch` and the request message has exactly one `FieldMask` message in it, additional code is rendered for the gateway handler that will populate the `FieldMask` based on the request body.

This handles two scenarios:
*The FieldMask is hidden from the REST request* (as in the `UpdateV2` example). In this case, the FieldMask is updated from the request body and set in the gRPC request message.

*The FieldMask is exposed to the REST request* (as in the `PatchWithFieldMaskInBody` example). For this case, a check is made as to whether the FieldMask is nil/empty prior to populating with the request body. If it's not nil, then it converts the FieldMask paths from the REST (snake_case) to gRPC (PascalCase) format. Otherwise, it acts like the previous case.

Additional comments of interest are inline on this PR.

### CODE REVIEW

File: protoc-gen-grpc-gateway/gengateway/template.go [hunk 1]
```diff
@@ -266,7 +306,11 @@ var (
 		grpclog.Infof("Failed to start streaming: %v", err)
 		return nil, metadata, err
 	}
-	dec := marshaler.NewDecoder(req.Body)
+	body, berr := ioutil.ReadAll(req.Body)
```
Comment (dmacthedestroyer) [inline protoc-gen-grpc-gateway/gengateway/template.go]: I had to add this to the `bidi-streaming-request-func` because I added the `ioutil` and `bytes` imports in the generator code above. Certain cases would cause compile errors for unused imports.

File: protoc-gen-grpc-gateway/gengateway/template.go [hunk 2]
```diff
@@ -210,9 +235,24 @@ var (
 	var protoReq {{.Method.RequestType.GoType .Method.Service.File.GoPkg.Path}}
 	var metadata runtime.ServerMetadata
 {{if .Body}}
-	if err := marshaler.NewDecoder(req.Body).Decode(&{{.Body.AssignableExpr "protoReq"}}); err != nil && err != io.EOF  {
+	body, berr := ioutil.ReadAll(req.Body)
```
Comment (dmacthedestroyer) [inline protoc-gen-grpc-gateway/gengateway/template.go]: I'd be glad to receive recommendations on how to do this in a way that doesn't require reading the entire request body multiple times. I wrote it like this because I thought it was the most readable.

File: runtime/fieldmask.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	fm := &field_mask.FieldMask{}
+	var root interface{}
+	if err := json.NewDecoder(r).Decode(&root); err != nil {
+		if err == io.EOF {
+			return fm, nil
+		}
+		return nil, err
+	}
+
+	queue := []struct {
```
Comment (achew22) [inline runtime/fieldmask.go]: Can you extract the struct into a named type (since you use it below) but leave it unexported

File: protoc-gen-grpc-gateway/gengateway/template.go [hunk 4]
```diff
@@ -54,6 +55,26 @@ func (b binding) QueryParamFilter() queryParamFilter {
 	return queryParamFilter{utilities.NewDoubleArray(seqs)}
 }
 
+// FieldMaskField returns the golang-style name of the variable for a FieldMask, if there is exactly one of that type in
+// the message. Otherwise, it returns an empty string.
+func (b binding) FieldMaskField() string {
+	var fieldMaskField *descriptor.Field
+	for _, f := range b.Method.RequestType.Fields {
+		if f.GetTypeName() == ".google.protobuf.FieldMask" {
```
[the remark this answers is not in the document]
Comment (dmacthedestroyer) [reply protoc-gen-grpc-gateway/gengateway/template.go]: I think I'll add another proto file and template tests to verify each of these boundary conditions

File: protoc-gen-grpc-gateway/gengateway/template.go [hunk 5]
```diff
@@ -54,6 +55,26 @@ func (b binding) QueryParamFilter() queryParamFilter {
...
+			// if there is more than 1 FieldMask for this request, then return none
```
[the remark this answers is not in the document]
Comment (dmacthedestroyer) [reply protoc-gen-grpc-gateway/gengateway/template.go]: I refrained from anything like that because I didn't want to assume that it's incorrect to have more than 1 fieldmask in a request proto. Also because I'm relatively new to Go templating and couldn't think of a more elegant solution :)

You could argue an error should be returned if none are found, as well, correct?

File: examples/server/a_bit_of_everything.go [hunk 6]
```diff
@@ -11,14 +11,15 @@ import (
...
-	sub2 "github.com/grpc-ecosystem/grpc-gateway/examples/proto/sub2"
+	"github.com/grpc-ecosystem/grpc-gateway/examples/proto/sub"
+	"github.com/grpc-ecosystem/grpc-gateway/examples/proto/sub2"
 	"github.com/rogpeppe/fastuuid"
 	"google.golang.org/genproto/googleapis/rpc/errdetails"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/metadata"
 	"google.golang.org/grpc/status"
+	"strings"
```
Comment (johanbrandhorst) [inline examples/server/a_bit_of_everything.go]: This should be at the top with the other standard library imports

**Comment by johanbrandhorst** [comment]
This is a great contribution! Could you please squash the commits and make sure there's a single commit with a new title and description? It'll help us when writing release notes as well as for historical purposes of course.

**Comment by achew22** [comment]
@johanbrandhorst when you click the merge will request button on GitHub there should be an option to squash the commits into 1 if you prefer that. That way you don't have to ask the people to squash them at the very end.

**Comment by achew22** [comment]
Unfortunately this PR has a merge conflict. Do you think you could run `make examples` one more time and upload them? Sorry about that! @johanbrandhorst feel free to squash this if you get to it before me.

**Comment by dmacthedestroyer** [comment]
This PR isn't quite ready yet... there are some broken unit tests and the build needs to be fixed. I'm still working on it but my availability has changed somewhat. Is it ok if I maintain this PR for a while and ping you guys when it's ready?

**Comment by achew22** [comment]
Sounds great to me! @mention me when you're ready and I'll take another look

**Comment by kibertoad** [comment]
Yay! Thanks for making this :)

**Comment by achew22** [comment]
Did those commits come from a merge? Is there any chance you could rebase off master?

**Comment by dmacthedestroyer** [comment]
yep, good call. @razamiDev took over work on this project a little while ago, thus his additional commits. Will rebase and then fix the broken builds

**Comment by razamiDev** [comment]
I am okay with my commits to be contributed to this project. My latest commit is an attempt to fix the travis.cli errors by updating the conflicting files.

**Comment by dmacthedestroyer** [comment]
@achew22 or @johanbrandhorst wanna give this CLA the OK so we can get the build fixed?

**Comment by achew22** [comment]
To my eyes, it looks like the Ci is also failing. Could you get that squared up first?

**Comment by dmacthedestroyer** [comment]
yep, working on learning what a Bazel is, then will try to fix that :)

**Comment by ivucica** [comment]
The errors are not due to Bazel, so I'd take a look at other failures first. From what I can see, you did not add `field_mask.proto` to third_party?

File: Makefile [hunk 7]
```diff
@@ -108,7 +108,7 @@ UNANNOTATED_ECHO_EXAMPLE_SRCS=$(EXAMPLE_CLIENT_DIR)/unannotatedecho/api_client.g
 EXAMPLE_CLIENT_SRCS=$(ECHO_EXAMPLE_SRCS) $(ABE_EXAMPLE_SRCS) $(UNANNOTATED_ECHO_EXAMPLE_SRCS)
 SWAGGER_CODEGEN=swagger-codegen
 
-PROTOC_INC_PATH=$(dir $(shell which protoc))/../include
+PROTOC_INC_PATH=$(dir $(shell which protoc))/../include
```
Comment (ivucica) [inline Makefile]: I'm not a fan of changes like this. Can you amend the commit (instead of adding to PR) to eliminate this change? Or if the intention is to squash the PR at the end, please do just remove this change.

**Comment by dmacthedestroyer** [comment]
hmm.. so I've got two issues with the build, it seems.

[`go vet` fails](https://travis-ci.org/grpc-ecosystem/grpc-gateway/jobs/415935077#L756), and fixing it results in [failed integration tests](https://github.com/grpc-ecosystem/grpc-gateway/blob/master/examples/integration/integration_test.go#L844)
_I'm not sure how this got through the gated checkin... I don't think this is anything I've modified. I can fix the lint issue, but will want some feedback on what the desired behavior should be._

It appears that the way in which my machine generates `.pb.go` files differs from the build agents in how it resolves the `FieldMask` message.
_I've seen this issue before, but forgot the solution. I can look into it further._

**Comment by johanbrandhorst** [comment]
@dmacthedestroyer check you've got the same version of `protoc` and `protoc-gen-go`. The `protoc`  version used is [`3.1.0`](https://github.com/grpc-ecosystem/grpc-gateway/blob/master/.travis.yml#L74) (holy hell that's old we should upgrade), and the `protoc-gen-go` version is [`b4deda0973fb4c70b50d226b1af49f3da59f5265`](https://github.com/grpc-ecosystem/grpc-gateway/blob/master/Gopkg.toml#L36).

File: docs/_docs/usage.md [hunk 8]
```diff
@@ -187,7 +187,8 @@ This parameter can be useful to pass request scoped context between the gateway
 * [How gRPC error codes map to HTTP status codes in the response](https://github.com/grpc-ecosystem/grpc-gateway/blob/master/runtime/errors.go#L15)
 * HTTP request source IP is added as `X-Forwarded-For` gRPC request header
 * HTTP request host is added as `X-Forwarded-Host` gRPC request header
-* HTTP `Authorization` header is added as `authorization` gRPC request header
+* HTTP `Authorization` header is added as `authorization` gRPC request header 
 * Remaining Permanent HTTP header keys (as specified by the IANA [here](http://www.iana.org/assignments/message-headers/message-headers.xhtml) are prefixed with `grpcgateway-` and added with their values to gRPC request header
-* HTTP headers that start with 'Grpc-Metadata-' are mapped to gRPC metadata (after removing prefix 'Grpc-Metadata-')
+* HTTP headers that start with 'Grpc-Metadata-' are mapped to gRPC metadata (prefixed with `grpcgateway-`)
```
Comment (johanbrandhorst) [inline docs/_docs/usage.md]: Is there a reason we changed this?

File: examples/server/a_bit_of_everything.go [hunk 9]
```diff
@@ -6,13 +6,15 @@ import (
 	"io"
 	"sync"
 
+	"strings"
```
Comment (johanbrandhorst) [inline examples/server/a_bit_of_everything.go]: Put this import with other stdlib imports.

File: protoc-gen-grpc-gateway/BUILD.bazel [hunk 10]
```diff
@@ -20,7 +20,6 @@ go_library(
 go_binary(
     name = "protoc-gen-grpc-gateway",
     embed = [":go_default_library"],
-    visibility = ["//visibility:public"],
```
Comment (johanbrandhorst) [inline protoc-gen-grpc-gateway/BUILD.bazel]: Why this change?

File: runtime/errors_test.go [hunk 11]
```diff
@@ -1,32 +1,27 @@
...
 import (
-	"context"
 	"encoding/json"
 	"fmt"
 	"net/http"
 	"net/http/httptest"
 	"strings"
 	"testing"
 
+	"context"
```
Comment (johanbrandhorst) [inline runtime/errors_test.go]: Put this import with the other stdlib imports

File: runtime/errors_test.go [hunk 12]
```diff
@@ -68,20 +57,5 @@ func TestDefaultHTTPError(t *testing.T) {
...
-				continue
-			}
-			if len(details) != 1 {
-				t.Errorf(`len(body["details"]) = %v; want 1`, len(details))
-				continue
-			}
-			if details[0].(map[string]interface{})["@type"] != spec.details {
-				t.Errorf(`details.@type = %s; want %s`, details[0].(map[string]interface{})["@type"], spec.details)
-			}
-		}
```
Comment (johanbrandhorst) [inline runtime/errors_test.go]: Why are we removing these tests? I'm really confused by this diff.

File: runtime/fieldmask.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+		lowerCasedParts := strings.Split(path, ".")
+		var camelCasedParts []string
+		for _, part := range lowerCasedParts {
+			camelCasedParts = append(camelCasedParts, generator.CamelCase(part))
+		}
+		newPaths = append(newPaths, strings.Join(camelCasedParts, "."))
+	}
+
+	mask.Paths = newPaths
+}
```
Comment (johanbrandhorst) [inline runtime/fieldmask.go]: This might just be me... but personally I think the field mask should contain the proto definition names. It's the only real source of truth that we have for what a fields name really is. Anything else is language specific. If I'm writing a service that accepts both FieldMasks for gRPC clients and a gRPC Gateway, I expect the field masks to use the protobuf field names.

File: runtime/query.go [hunk 14]
```diff
@@ -221,23 +221,6 @@ func populateField(f reflect.Value, value string, props *proto.Properties) error
...
-			if err != nil {
-				return fmt.Errorf("bad Duration: %v", err)
-			}
-
-			ns := d.Nanoseconds()
-			s := ns / 1e9
-			ns %= 1e9
-			f.Field(0).SetInt(s)
-			f.Field(1).SetInt(ns)
-			return nil
```
Comment (johanbrandhorst) [inline runtime/query.go]: OK I'm starting to think there's something weird going on with my diff viewer. Do you need to rebase your commits on a recent master?

**Comment by johanbrandhorst** [comment]
Also, since we're getting ready to merge this, please squash the commits and add a nice explanatory commit message.

As for the vet failure, I'm not sure I understand? I looked at the diff but it's full of unrelated stuff.

File: runtime/query.go (hunk 14 above)
Comment (dmacthedestroyer) [reply runtime/query.go]: these weird unrelated diffs concern me as well. I just recently merged `master` into this feature branch, so I'm not sure why these diffs show up. I'll try to figure this out.

**Comment by dmacthedestroyer** [comment]
I'll ignore the vet failure for now, since it seems I've got bigger problems to figure out first :)

As for the squash commit, how would you recommend I do that within this PR? I've always in the past squashed the commits upon merging the PR.

File: runtime/fieldmask.go (hunk 13 above)
Comment (dmacthedestroyer) [reply runtime/fieldmask.go]: hmm... that's interesting. I think I intuitively did it this way because I had reflection on my mind. 

I see your point, though, and I agree. Will change.

File: docs/_docs/usage.md (hunk 8 above)
Comment (razamiDev) [reply docs/_docs/usage.md]: @johanbrandhorst this change came from merging with master. The current repo has it like this.

**Comment by ivucica** [comment]
@dmacthedestroyer If you want to squash commits yourself, `git rebase -i` and then mark individual commits you want to squash with their parents with `f` (standing for`fixup`).

You won't be able to push the result as you usually would (as you just rewrote history) but `git push -f` is a way out.

Alternatively, if @dmacthedestroyer agrees, @johanbrandhorst how about we just squash using Github UI (which will use PR description as the commit message)? Maybe that's better. Not sure.

---

What we *do* separately *need*, however, is a rebase on top of current master. I always have to look up how to do that, but `git rebase` is certainly involved, as is `git push -f` (because you're rewriting history).

**Comment by johanbrandhorst** [comment]
I believe `git rebase -i origin/master` will interactively rebase the current branch on top of current master, which will allow both squashing and rebasing at once.

Then `git push -f yourremote yourbranch` should fix the commits on the remote.

**Comment by dmacthedestroyer** [comment]
my current, bigger problem is hot to get this build to pass... What's the standard for properly importing `FieldMask` in proto files for Go? I'm assuming if I add [this file](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/field_mask.proto) somewhere in the project and reference it properly, it should give the proper import path. Is this how people are expected to incorporate `FieldMask` into their projects?

**Comment by ivucica** [comment]
Yes, as a developer defining the `field_mask.proto`-aware API, you are meant to:

    import "google/protobuf/field_mask.proto";

Then, the developer should not need to `protoc` it because, as it is a WKT (well known type). They can and should just import the pregenerated .pb.go should you need to instantiate the generated Go struct type:

    import (
        "google.golang.org/genproto/protobuf/field_mask"
    )

Pretty much as you have done in the examples you updated/created.

I am, off the top of my head, not sure how to update Bazel files. Probably inspiration can be drawn from other WKTs that may be used such as duration or timestamp.

**Comment by dmacthedestroyer** [comment]
I ran into and addressed another issue with unused imports after merging the latest master.
My remaining tasks for this PR:

* include the `field_mask.proto` file into the project to fix the build
* change field mask logic to use proto definition names instead of Golang names (from review comments)
* fix edge case issue with how I'm allocating arrays in `runtime/fieldmask.go` (found in another issue)

I've lost my corporate support for this issue so I'm going to tackle finishing it up on my own nights & weekends. I plan to push ahead to get it ready for merging next week.

Sorry it's taken so long :(

**Comment by johanbrandhorst** [comment]
All the best @dmacthedestroyer - we're all volunteers here :). Let me know if there's anyway we can help you!

**Comment by johanbrandhorst** [comment]
We've release v1.5.0 without this, but we're still very keen on having it! Lets us know if there's anything we can do to help.

**Comment by vtolstov** [comment]
@dmacthedestroyer hi, do you plan to complete pr ?

**Comment by razamiDev** [comment]
@vtolstov @johanbrandhorst I will try to complete this item towards the end of this week or beginning of next. Me and @dmacthedestroyer have both worked on this and I have taken it over.

**Comment by johanbrandhorst** [comment]
My thinking is we just check it into the third_party folder.

**Comment by johanbrandhorst** [comment]
Actually, field_mask.proto is a well known type, it shouldn't need to be checked in. Whatever we do to get it building wity empty.proto and duration.proto, we just need to do the same thing as there.

**Comment by johanbrandhorst** [comment]
Please rebase on master for new CI functionality, regenerating should be much easier now!

File: docs/_docs/usage.md (hunk 8 above)
Comment (ivucica) [reply docs/_docs/usage.md]: If a change is in the master, then, to my understanding, there should be no diff visible.

File: docs/_docs/usage.md [hunk 15]
```diff
@@ -187,7 +187,8 @@ This parameter can be useful to pass request scoped context between the gateway
 * [How gRPC error codes map to HTTP status codes in the response](https://github.com/grpc-ecosystem/grpc-gateway/blob/master/runtime/errors.go#L15)
 * HTTP request source IP is added as `X-Forwarded-For` gRPC request header
 * HTTP request host is added as `X-Forwarded-Host` gRPC request header
-* HTTP `Authorization` header is added as `authorization` gRPC request header
+* HTTP `Authorization` header is added as `authorization` gRPC request header
```
Comment (ivucica) [inline docs/_docs/usage.md]: I am not a fan of 'add unnecessary trailing space' changes ;-)

File: examples/clients/abe/camel_case_service_name_api.go [hunk 16]
```diff
@@ -37,8 +37,7 @@ func NewCamelCaseServiceNameApiWithBasePath(basePath string) *CamelCaseServiceNa
 }
 
 /**
- * Create a new ABitOfEverything
- * This API creates a new ABitOfEverything
+ *
```
Comment (ivucica) [inline examples/clients/abe/camel_case_service_name_api.go]: I know this is autogenerated, but I'm slightly confused.

File: protoc-gen-swagger/genswagger/template.go [hunk 17]
```diff
@@ -635,10 +634,9 @@ func renderServices(services []*descriptor.Service, paths swaggerPathsObject, re
 					}
 
 					parameters = append(parameters, swaggerParameterObject{
-						Name:        parameter.String(),
-						Description: desc,
```
Comment (ivucica) [inline protoc-gen-swagger/genswagger/template.go]: Why remove desc?

File: protoc-gen-swagger/genswagger/template.go [hunk 18]
```diff
@@ -662,15 +659,11 @@ func renderServices(services []*descriptor.Service, paths swaggerPathsObject, re
...
-						if schema.Description != "" {
-							desc = schema.Description
-						} else {
-							desc = fieldProtoComments(reg, lastField.Target.Message, lastField.Target)
-						}
 					}
 
+					desc := ""
 					if meth.GetClientStreaming() {
-						desc += " (streaming inputs)"
```
Comment (ivucica) [inline protoc-gen-swagger/genswagger/template.go]: This should have the space. This should have the +.

Come to think of it, this looks like it MIGHT be a rollback of a pull request I approved recently. Why wouldn't you want to have `fieldProtoComments` result used for desc? (This might be the reason for disappearing comments in the autogenerated code above.)

Same applies for many other changes in this file: they look like rollbacks. Please don't rollback needlessly and please rebase on master correctly without steamrolling older changes :-)

File: runtime/fieldmask.go (hunk 13 above)
Comment (ivucica) [reply runtime/fieldmask.go]: Was this done?

File: runtime/query.go (hunk 14 above)
Comment (ivucica) [reply runtime/query.go]: Instead of merging, you should be rebasing :-)

**Comment by johanbrandhorst** [comment]
Superceded by #806
